src/linker.rs

```diff
@@ -141,6 +137,58 @@ pub enum OptLevel {
     SizeMin,
 }
 
+pub struct FileInput<'a> {
+    path: &'a Path,
+    file: File,
+}
+
+pub struct BytesInput<'a> {
+    name: &'a str,
+    cursor: io::Cursor<&'a [u8]>,
+}
+
+pub enum LinkerInput<'a> {
+    File(FileInput<'a>),
+    Bytes(BytesInput<'a>),
+}
+
+impl<'a> Seek for LinkerInput<'a> {
+    fn seek(&mut self, pos: io::SeekFrom) -> io::Result<u64> {
+        match self {
+            LinkerInput::File(input) => input.file.seek(pos),
+            LinkerInput::Bytes(input) => input.cursor.seek(pos),
+        }
+    }
+}
+
+impl<'a> Read for LinkerInput<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self {
+            LinkerInput::File(input) => input.file.read(buf),
+            LinkerInput::Bytes(input) => input.cursor.read(buf),
+        }
+    }
+}
+
+impl<'a> From<(&'a str, &'a [u8])> for LinkerInput<'a> {
```

Comment:
This doesn't look great. Now that there are two ways to create LinkerInput, I'd
make it new_from_file and new_from_buffer